pkg/ccl/streamingccl/streamingest/stream_ingest_manager.go

```diff
@@ -52,14 +59,124 @@ func (r *streamIngestManagerImpl) GetStreamIngestionStats(
 func (r *streamIngestManagerImpl) GetReplicationStatsAndStatus(
 	ctx context.Context, ingestionJobID jobspb.JobID,
 ) (*streampb.StreamIngestionStats, string, error) {
 	return getReplicationStatsAndStatus(ctx, r.jobRegistry, r.txn, ingestionJobID)
 }
 
+// RevertTenantToTimestamp  implements streaming.StreamIngestManager interface.
+func (r *streamIngestManagerImpl) RevertTenantToTimestamp(
+	ctx context.Context, tenantName roachpb.TenantName, revertTo hlc.Timestamp,
+) error {
+	execCfg := r.evalCtx.Planner.ExecutorConfig().(*sql.ExecutorConfig)
+
+	// These vars are set in Txn below. This transaction checks
+	// the service state of the tenant record, moves the tenant's
+	// data state to ADD, and installs a PTS for the revert
+	// timestamp.
+	//
+	// NB: We do this using a different txn since we want to be
+	// able to commit the state change during the
+	// non-transactional RevertSpans below.
+	var (
+		originalDataState mtinfopb.TenantDataState
+		tenantID          roachpb.TenantID
+		ptsCleanup        func()
+	)
+	if err := execCfg.InternalDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		tenantRecord, err := sql.GetTenantRecordByName(ctx, r.evalCtx.Settings, r.txn, tenantName)
+		if err != nil {
+			return err
+		}
+		if tenantRecord.ServiceMode != mtinfopb.ServiceModeNone {
+			return errors.Newf("cannot revert tenant %q (%d) in service mode %s; service mode must be %s",
+				tenantRecord.Name,
+				tenantRecord.ID,
+				tenantRecord.ServiceMode,
+				mtinfopb.ServiceModeNone,
+			)
+		}
+
+		originalDataState = tenantRecord.DataState
+		tenantID, err = roachpb.MakeTenantID(tenantRecord.ID)
+		if err != nil {
+			return err
+		}
+		ptsCleanup, err = protectTenantSpanWithSession(ctx, r.evalCtx, txn, execCfg, tenantID, r.sessionID, revertTo)
+		if err != nil {
+			return errors.Wrap(err, "protecting revert timestamp")
+		}
+
+		// Set the data state to Add during the destructive operation.
+		//
+		// TODO(ssd): We likely also want to record the revert
+		// timestamp in the tenant metadata so that we have a
+		// record of the revert and so that we can use it to
+		// verify other operations that may want to start from
+		// that same timestamp.
+		tenantRecord.DataState = mtinfopb.DataStateAdd
+		return sql.UpdateTenantRecord(ctx, r.evalCtx.Settings, r.txn, tenantRecord)
+	}); err != nil {
+		return err
+	}
+	defer ptsCleanup()
+
+	spanToRevert := keys.MakeTenantSpan(tenantID)
+	if err := sql.RevertSpansFanout(ctx, r.evalCtx.Txn.DB(), r.evalCtx.JobExecContext.(sql.RevertSpansContext),
+		[]roachpb.Span{spanToRevert},
+		revertTo,
+		false, /* ignoreGCThreshold */
+		int64(sql.RevertTableDefaultBatchSize),
+		nil /* onCompletedCallback */); err != nil {
+		return err
+	}
+
+	return execCfg.InternalDB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		tenantRecord, err := sql.GetTenantRecordByName(ctx, r.evalCtx.Settings, r.txn, tenantName)
```

Comment:
you're passing in r.txn here, I think you meant to pass in txn instead?

Reply:
Good catch, thanks!